### PR TITLE
feat: replicate event bus kafka

### PR DESCRIPTION
**Pull request by navinkarkera**
**Description:** Replicates [event_bus_kafka](https://github.com/openedx/event-bus-kafka) and removes kafka specific code. Cleans up producer and consumer to make it easier to add redis streams implementation.

_Since there are a lot of files added, it would be easier to review commit by commit._

**Tickets:** 
* `Private-ref`: [BB-7169](https://tasks.opencraft.com/browse/BB-7169)
* #1 

**Merge deadline:** List merge deadline (if any)

**Testing instructions:**

1. Make sure tests are passing.
2. Make sure kafka specific code is removed.

**Reviewers:**
- [ ] @keithgg
- [ ] @bmtcril 

**Merge checklist:**
- [ ] All reviewers approved
- [ ] CI build is green
- [ ] Version bumped
- [ ] Changelog record added
- [ ] Documentation updated (not only docstrings)
- [ ] Commits are squashed

**Post merge:**
- [ ] Create a tag
- [ ] Check new version is pushed to PyPI after tag-triggered build is 
      finished.
- [ ] Delete working branch (if not needed anymore)

### CODE REVIEW

**Comment by openedx-webhooks** [comment]
Thanks for the pull request, @navinkarkera!  Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket as you can:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


Please let us know once your PR is ready for our review and all tests are green.

File: edx_event_bus_redis/internal/consumer.py [hunk 1]
```diff
@@ -0,0 +1,557 @@
...
+                    if msg is None:
+                        time.sleep(POLL_FAILURE_SLEEP)
+                # if msg:
+                    # theoretically we could just call consumer.commit() without passing the specific message
+                    # to commit all this consumer's current offset across all partitions since we only process one
+                    # message at a time, but limit it to just the offset/partition of the specified message
+                    # to be super safe
+                    # self.consumer.commit(message=msg)
+        finally:
+            # TODO: close consumber
```
Comment (bmtcril) [inline edx_event_bus_redis/internal/consumer.py:225]:
```suggestion
            # TODO: close consumer
```

File: edx_event_bus_redis/internal/producer.py [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+        that can be used to recreate them.
+
+        Arguments:
+            err: Error if event production failed
+            evt: Event that was delivered (or failed to be delivered)
+        """
+        if err is not None:
+            record_producing_error(err, self)
+        else:
+            # Audit logging on success. See ADR: docs/decisions/0010_audit_logging.rst
```
Comment (bmtcril) [inline edx_event_bus_redis/internal/producer.py:71]: We should expand this link to go back to the Kafka docs, I think. In fact, as part of this work we should move any ADRs from the Kafka repo to openedx-events as more generally useful.

File: edx_event_bus_redis/internal/producer.py [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+            if not AUDIT_LOGGING_ENABLED.is_enabled():
+                return
+
+            # `evt.headers()` is None in this callback, so we need to use the bound data.
```
Comment (bmtcril) [inline edx_event_bus_redis/internal/producer.py:75]: I think this comment is not needed here?

File: edx_event_bus_redis/internal/consumer.py [hunk 4]
```diff
@@ -0,0 +1,557 @@
...
+        CONSECUTIVE_ERRORS_LIMIT = getattr(settings, 'EVENT_BUS_REDIS_CONSUMER_CONSECUTIVE_ERRORS_LIMIT', None)
+
+        try:
+            full_topic = get_full_topic(self.topic)
+            run_context = {
+                'full_topic': full_topic,
+                'consumer_group': self.group_id,
+                'expected_signal': self.signal,
+            }
+            # TODO: Consume events
```
Comment (bmtcril) [inline edx_event_bus_redis/internal/consumer.py:171]:
```suggestion

```

I think this is consuming events now?

**Review by bmtcril — APPROVED** [review]
This is a great start! I'm leaving some notes that are mostly nits on comments and things, but I think I'm going to merge this as-is to get the repo bootstrapped with some actual starting code and make sure you're unblocked. If you can address the comments in a follow-on PR that would be great.

**Comment by openedx-webhooks** [comment]
@navinkarkera 🎉 Your pull request was merged!  Please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/event-bus-redis&entry.867055334=https://github.com/openedx/event-bus-redis/pull/3&entry.1484655318=https://github.com/navinkarkera&entry.752974735=2023-02-27+15:29&entry.1917517419=2023-02-27+20:56&entry.2133058324=Yes) so we can improve your experience in the future.

File: edx_event_bus_redis/internal/__init__.py [hunk 5]
```diff
@@ -0,0 +1,7 @@
+"""
+Most of the implementation of the package is here and is internal-only.
+
+Public API will be in the ``edx_event_bus_redis`` module for the most part.
+
+See ADR ``docs/decisions/0006-public-api-and-app-organization.rst`` for the reasoning.
```
[the remark this answers is not in the document]
Comment (navinkarkera) [reply edx_event_bus_redis/internal/__init__.py:6]: Actually, I missed this comment, but it make sense to move it to openedx-events as both implementation will follow the same design.

File: edx_event_bus_redis/internal/consumer.py [hunk 6]
```diff
@@ -0,0 +1,557 @@
...
+
+
+class RedisEventConsumer:
+    """
+    Construct consumer for the given topic, group, and signal. The consumer can then
+    emit events from the event bus using the configured signal.
+
+    Note that the topic should be specified here *without* the optional environment prefix.
+
+    Can also consume messages indefinitely off the queue.
```
[the remark this answers is not in the document]
Comment (navinkarkera) [reply edx_event_bus_redis/internal/consumer.py:98]: I'll update these during implementation of the consumer.

File: edx_event_bus_redis/internal/producer.py (hunk 3 above)
Comment (navinkarkera) [reply edx_event_bus_redis/internal/producer.py:75]: Yep, this whole function might change as redis streams don't have a delivery callback. It returns a msgid immediately after adding an item.
```python
msgid = stream.add({'message': 'hello streams'})
print(msgid)

# Prints something like:
# b'1539008591844-0'
```

File: edx_event_bus_redis/internal/consumer.py (hunk 4 above)
Comment (navinkarkera) [reply edx_event_bus_redis/internal/consumer.py:171]: The actual work of fetching events from stream needs to be implemented. Actually the whole function would need to be updated to work with redis streams and consumer groups. I'll create separate tickets for producer and consumer implementation.

**Pull request merged**